File: TimeFlies/vc/done/DoneTableViewController.m
//
//  DoneTableViewController.m
//  testTabbarController
//
//  Created by 林伟池 on 15/10/26.
//  Copyright © 2015年 林伟池. All rights reserved.
//

#import "DoneTableViewController.h"
#import "WeixinShare.h"
#import "DoneTableViewCell.h"
#import "DoneDetailController.h"
#import "DoneModel.h"

@interface DoneTableViewController ()

@end

#define const_open_done_deteail @"open_done_detail_board"

@implementation DoneTableViewController
{
    long myIndex;
}
- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
    [[NSNotificationCenter defaultCenter] addObserverForName:@"DoneModelChange" object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        [self.tableView reloadData];
    }];
    
    if (![WXApi isWXAppInstalled]) {
        self.navigationItem.rightBarButtonItem = nil;
    }
    
    [[NSNotificationCenter defaultCenter] addObserverForName:@"FavoriteSuccess" object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        [[DoneModel instance] clearCache];
    }];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
}


- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:const_open_done_deteail]) {
        DoneDetailController* controller = segue.destinationViewController;
        controller.myIndex = myIndex;
    }
}

- (IBAction)onSave:(id)sender
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"分享到微信收藏", nil) message:NSLocalizedString(@"把时间轴所有的信息收藏到微信，同时清除时间轴的信息", nil) preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"否" style:UIAlertActionStyleDefault handler:nil];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"是" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        
        
        NSString* str = @"";
        for (int i = 0; i < [[DoneModel instance] getDoneCounts]; ++i) {
            PassThing* item = [[DoneModel instance] getThingByIndex:i];
            if (item) {
                NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
                [dateFormatter setDateFormat:NSLocalizedString(@"yyyy年MM月dd日", nil)];
                
                str = [NSString stringWithFormat:@"%@%@ -- %@\n%@\n", str, [dateFormatter stringFromDate:item.startTime], [dateFormatter stringFromDate:item.endTime], item.text];
            }
        }
        
        [[WeixinShare instance] sendTextContent:str Scene:WXSceneFavorite];
        
    }];
    [alertController addAction:okAction];
    [alertController addAction:cancelAction];
    
    [self presentViewController:alertController animated:YES completion:nil];
}
#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
#warning Incomplete implementation, return the number of sections
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
#warning Incomplete implementation, return the number of rows
    return [[DoneModel instance] getDoneCounts];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    DoneTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"done" forIndexPath:indexPath];
    
    PassThing* item = [[DoneModel instance] getThingByIndex:indexPath.row];
    if (item) {
        [cell viewInitWithTime:item.endTime Text:item.text];
    }
    else{
        [cell clear];
    }
    
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section
{
    return 0.01;
}

- (NSIndexPath *)tableView:(UITableView *)tableView willSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    myIndex = indexPath.row;
    [self performSegueWithIdentifier:const_open_done_deteail sender:self];
    return nil;
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: TimeFlies/vc/done/DoneTableViewController.swift
//
//  DoneTableViewController.swift
//  TimeFlies
//
//  Created by 林伟池 on 16/4/8.
//  Copyright © 2016年 林伟池. All rights reserved.
//

import Foundation

class SwiftDoneTableViewController: UITableViewController {
    
    //MARK: property
    let const_open_done_detail = "open_done_detail_board"
    var myIndex:Int = 0
    
    
    
    //MARK: init
    override func viewDidLoad() {
        super.viewDidLoad()
        
        NotificationCenter.default.addObserver(forName: NSNotification.Name(rawValue: "DoneModelChange"), object: nil, queue: nil) { [unowned self](note) -> Void in
            self.tableView.reloadData()
        }
        
        if !WXApi.isWXAppInstalled() {
            navigationItem.rightBarButtonItem = nil
        }

        NotificationCenter.default.addObserver(forName: NSNotification.Name(rawValue: "FavoriteSuccess"), object: nil, queue: nil) { (note) -> Void in
            DoneModel.instance().clearCache()
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == const_open_done_detail {
            let controller = segue.destination as! SwiftDoneDetailController
            controller.myIndex = myIndex
        }
    }
    //MARK: ui
    @IBAction func onSave(_ sender: NSObject) {
        let alertController = UIAlertController(title: NSLocalizedString("分享到微信收藏", comment: lyCommentDefault), message: NSLocalizedString("把时间轴所有的信息收藏到微信，同时清除时间轴的信息", comment: lyCommentDefault), preferredStyle: .alert)
        
        let cancelAction = UIAlertAction(title: NSLocalizedString("否", comment: lyCommentDefault), style: .default, handler: nil)
        let okAction = UIAlertAction(title: NSLocalizedString("是", comment: lyCommentDefault), style: .default) { (action) -> Void in
            var str:String = ""
            
            for i in 0 ..< DoneModel.instance().getDoneCounts() {
                let item = DoneModel.instance().getThingBy(i)
                let dateformatter = DateFormatter()
                dateformatter.dateFormat = NSLocalizedString("yyyy年MM月dd日", comment: lyCommentDefault)
                
                str = String(format: "%@%@ -- %@\n%@\n", str, dateformatter.string(from: (item?.startTime)!), dateformatter.string(from: (item?.endTime)!), (item?.text)!)
            }
            
            WeixinShare.instance().sendTextContent(str, scene: WXSceneFavorite)
        }
        
        alertController.addAction(okAction)
        alertController.addAction(cancelAction)
        
        present(alertController, animated: true, completion: nil)
    }
    
    @IBAction func onSwipeRight(_ sender: NSObject) {
        tabBarController?.selectedIndex = 2
    }
    
    @IBAction func onSwipeLeft(_ sender: NSObject) {
        
    }
    //MARK: delegate
    
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return DoneModel.instance().getDoneCounts()
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "done", for: indexPath) as! SwiftDoneTableViewCell
        
        let item = DoneModel.instance().getThingBy(indexPath.row)
        if item != nil {
            cell.viewInitWithTime((item?.endTime)!, text: (item?.text)!)
        }
        else {
            cell.clear()
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 0.01
    }
    
    override func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
        myIndex = indexPath.row
        performSegue(withIdentifier: const_open_done_detail, sender: nil)
        
        return nil
    }
    
    //MARK: other
    
}
